packages/loopring_v3/contracts/impl/libexchange/ExchangeWithdrawals.sol

```diff
@@ -283,44 +284,56 @@ library ExchangeWithdrawals
-            data = data & uint(~((1 << (7 * 8)) - 1));
-
-            // Update the data in storage
-            if (withdrawBlock.withdrawals.length >= 32) {
-                assembly {
-                    mstore(add(slice, offset), data)
-                }
-                uint dataSlot1 = slice[0];
-                uint dataSlot2 = slice[1];
-                assembly {
-                    sstore(add(sc, slot1), dataSlot1)
-                    sstore(add(sc, slot2), dataSlot2)
-                }
+            address recipient = S.accounts[accountID].owner;
+            // Transfer the tokens
+            if (useGasLimit) {
+                success = sendTokens(
+                    S,
+                    recipient,
+                    tokenID,
+                    amount
+                );
             } else {
-                bytes memory mWithdrawals = withdrawBlock.withdrawals;
-                assembly {
-                    mstore(add(mWithdrawals, offset), data)
-                }
-                withdrawBlock.withdrawals = mWithdrawals;
+                success = transferTokens(
+                    S,
+                    recipient,
+                    tokenID,
+                    amount
+                );
             }
 
-            ExchangeData.Account storage account = S.accounts[accountID];
-
-            // Transfer the tokens
-            withdrawTokens(
-                S,
-                account.owner,
-                tokenID,
-                amount
-            );
+            if (success) {
+                // Set everything to 0 for this withdrawal so it cannot be used anymore
+                data = data & uint(~((1 << (7 * 8)) - 1));
+
+                // Update the data in storage
+                if (withdrawBlock.withdrawals.length >= 32) {
+                    assembly {
+                        mstore(add(slice, offset), data)
+                    }
+                    uint dataSlot1 = slice[0];
+                    uint dataSlot2 = slice[1];
+                    assembly {
+                        sstore(add(sc, slot), dataSlot1)
+                        sstore(add(sc, add(slot, 1)), dataSlot2)
+                    }
+                } else {
+                    bytes memory mWithdrawals = withdrawBlock.withdrawals;
+                    assembly {
+                        mstore(add(mWithdrawals, offset), data)
+                    }
+                    withdrawBlock.withdrawals = mWithdrawals;
+                }
 
-            emit WithdrawalCompleted(
-                accountID,
-                tokenID,
-                account.owner,
-                uint96(amount)
-            );
+                emit WithdrawalCompleted(
+                    accountID,
+                    tokenID,
+                    recipient,
+                    uint96(amount)
+                );
+            }
+        } else {
+            success = true;
```

Comment:
should we also emit WithdrawalCompleted event if the amount is zero?

Reply:
WithdrawalCompleted is now emitted when the withdrawal data hasn't been reset yet by a previous successful withdrawal (either the account, token or amount is non-zero). Otherwise we don't have any actual information about the withdrawal onchain.